File: python/160.py
class Solution:
    # @param headA: the first list
    # @param headB: the second list
    # @return: a ListNode
    def getIntersectionNode(self, headA, headB):
        # Write your code here
        listA = []
        listB = []
        if headA == None or headB == None:
            return None
        while 1:
            if headA == None:
                break;
            listA.append(headA.val)
            headA = headA.next

        while 1:
            if headB == None:
                break;
            listB.append(headB.val)
            headB = headB.next

        if listA[-1] != listB[-1]:
            return None;
        
        if len(listA)<len(listB):
            minLen = len(listA)
        else:
            minLen = len(listB)

        inster = []
        for i in range(1,minLen+1):
            if listA[-i] != listB[-i]:
                return ListNode(listA[-i+1])
            if i == minLen:
                return ListNode(listA[-i])


#not best
class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if headA is None or headB is None:
            return None

        pa = headA # 2 pointers
        pb = headB

        while pa is not pb:
            # if either pointer hits the end, switch head and continue the second traversal, 
            # if not hit the end, just move on to next
            pa = headB if pa is None else pa.next
            pb = headA if pb is None else pb.next

        return pa 

File: c++/160.cpp
/**
 * Definition for singly-linked list.
 */
 struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(nullptr) {}
 };

int get_list_length(ListNode *head) {
    int len = 0;
    while (head) {
        len++;
        head = head->next;
    }
    return len;
}

ListNode *forward_long_list(int long_len,int short_len,ListNode *head) {
    int delta = long_len - short_len;
    while (head && delta) {
        head = head->next;
        delta--;
    }
    return head;
}

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int list_a_len = get_list_length(headA);
        int list_b_len = get_list_length(headB);
        if (list_a_len > list_b_len) {
            headA = forward_long_list(list_a_len,list_b_len,headA);
        } else {
            headB = forward_long_list(list_b_len,list_a_len,headB);
        }
        while (headA && headB) {
            if (headB == headA) {
                return headA;
            }
            headA = headA->next;
            headB = headB->next;            
        }
        return nullptr;
    }
};